jpa-criteria/impl/src/main/java/com/blazebit/persistence/criteria/impl/InternalSetQuery.java

```diff
@@ -0,0 +1,194 @@
+import com.blazebit.persistence.criteria.BlazeSubquery;
+import com.blazebit.persistence.criteria.impl.RenderContext.ClauseType;
+import com.blazebit.persistence.criteria.impl.expression.AbstractSelection;
+import com.blazebit.persistence.criteria.impl.expression.SubqueryExpression;
+import com.blazebit.persistence.criteria.impl.path.AbstractFrom;
+import com.blazebit.persistence.criteria.impl.path.AbstractJoin;
+import com.blazebit.persistence.criteria.impl.path.EntityJoin;
+import com.blazebit.persistence.criteria.impl.path.RootImpl;
+import com.blazebit.persistence.criteria.impl.path.TreatedPath;
+
+import javax.persistence.Tuple;
+import javax.persistence.criteria.Expression;
+import javax.persistence.criteria.Join;
+import javax.persistence.criteria.Order;
+import javax.persistence.criteria.ParameterExpression;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import javax.persistence.criteria.Selection;
+import javax.persistence.criteria.Subquery;
+import javax.persistence.metamodel.Attribute;
+import javax.persistence.metamodel.EntityType;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * @author Christian Beikov
+ * @since 1.2.0
+ */
+public class InternalSetQuery<T> implements Serializable {
+
+    private static final long serialVersionUID = 1L;
+
+    private final BlazeCriteriaSetQuery<T> owner;
+    private final BlazeCriteriaBuilderImpl criteriaBuilder;
+
+    private final BlazeCriteriaQueryImpl<T> lhs, rhs;
+    private List<BlazeOrder> orderList = Collections.emptyList();
+
+    public InternalSetQuery(BlazeCriteriaSetQuery<T> owner, BlazeCriteriaBuilderImpl criteriaBuilder, BlazeCriteriaQueryImpl<T> lhs, BlazeCriteriaQueryImpl<T> rhs) {
+        this.owner = owner;
+        this.criteriaBuilder = criteriaBuilder;
+        this.lhs = lhs;
+        this.rhs = rhs;
+    }
+
+    /* Order by */
+
+    public List<BlazeOrder> getBlazeOrderList() {
+        return orderList;
+    }
+
+    public void setBlazeOrderList(List<BlazeOrder> orderList) {
+        this.orderList = orderList;
+    }
+
+    @SuppressWarnings({"unchecked"})
+    public List<Order> getOrderList() {
+        return (List<Order>) (List<?>) orderList;
+    }
+
+    @SuppressWarnings({"unchecked"})
+    public void setOrderList(List<Order> orderList) {
+        this.orderList = (List<BlazeOrder>) (List<?>) orderList;
+    }
+
+    /* Parameters */
+
+    public Set<ParameterExpression<?>> getParameters() {
+        // NOTE: we have to always visit them because it's not possible to cache that easily
+        ParameterVisitor visitor = new ParameterVisitor();
+
+        return visitor.getParameters();
+    }
+
+    /* Rendering */
+
+    public FinalSetOperationCriteriaBuilder<T> render(CriteriaBuilder<T> cb) {
```

Comment:
Left-nested sets actually have to be rendered against a `StartOngoingSetOperationCriteriaBuilder`.

Reply:
I guess we will have to use Object here as return type then as well, or maybe some common interface if you find one.